### PR TITLE
Updated code to support the IDocument interface

### PR DESCRIPTION
**Pull request by Pauljgraham**

### CODE REVIEW

File: SFA.DAS.CosmosDb/SFA.DAS.CosmosDb/DocumentRepository.cs [hunk 1]
```diff
@@ -21,22 +21,24 @@ protected DocumentRepository(IDocumentClient documentClient, string databaseName
             _collectionName = collectionName;
         }
 
-        public Task Add(TDocument document, RequestOptions requestOptions = null, CancellationToken cancellationToken = default)
+        public virtual Task Add(TDocument document, RequestOptions requestOptions = null, CancellationToken cancellationToken = default)
         {
-            return _documentClient.CreateDocumentAsync(UriFactory.CreateDocumentCollectionUri(_databaseName, _collectionName), document, requestOptions, false, cancellationToken);
+            if (document == null) throw new ArgumentNullException(nameof(document));
+            document.Id = document.Id != Guid.Empty ? document.Id : Guid.NewGuid();
```
Comment (elliotritchie) [inline SFA.DAS.CosmosDb/SFA.DAS.CosmosDb/DocumentRepository.cs]: Does `Id` need to be set here?

Comment (Pauljgraham) [reply SFA.DAS.CosmosDb/SFA.DAS.CosmosDb/DocumentRepository.cs]: Oddly yes it does, if we pass in an Empty Id it uses it as the DocumentId (regardless of the true/false) value.

Comment (elliotritchie) [reply SFA.DAS.CosmosDb/SFA.DAS.CosmosDb/DocumentRepository.cs]: Ah I see.

If it always has to have one what do you think to throwing if it's empty instead, like you do in `Update`? That way the `Id` property can be encapsulated by the document i.e.

```
public interface IDocument
{
    Guid Id { get; } // No set
    string ETag { get; } // No set
}
```

Comment (Pauljgraham) [reply SFA.DAS.CosmosDb/SFA.DAS.CosmosDb/DocumentRepository.cs]: Doing this, seems to code more unclear. If the Guid is not set by the caller and given that cosmos will insert an Empty Guid into the document. We then need to create allow the IDocument to also have a set property. 

```c#
public interface IDocument
{
    Guid Id { get;  set; } // No set
    string ETag { get; } // No set
}
```
But this then looks odd, so making the interface more open seemed like the best option.

Comment (Pauljgraham) [reply SFA.DAS.CosmosDb/SFA.DAS.CosmosDb/DocumentRepository.cs]: We can implement your suggestion, by stating that the class must generate the Id.

**Pull request merged**